common/include/pcl/common/impl/eigen.hpp

```diff
@@ -248,6 +249,39 @@ pcl::computeCorrespondingEigenVector (const Matrix& mat, const typename Matrix::
     eigenvector = vec3 / std::sqrt (len3);
 }
 
+namespace pcl {
+namespace detail{
+template <typename Vector, typename Scalar>
+struct EigenVector {
+  Vector vector;
+  Scalar length;
+};
+
+/**
+ * @brief returns the unit vector along the largest eigen value as well as the
+ *        length of the largest eigenvector
+ * @tparam Matrix matrix type providing similar in API as Eigen::Matrix
+ */
+template <typename Vector, typename Matrix> static EigenVector<Vector, typename Matrix::Scalar>
+get_largest_3x3_eigenvector (const Matrix scaledMatrix)
+{
+  Matrix crossProduct;
+  crossProduct << scaledMatrix.row (0).cross (scaledMatrix.row (1)),
+                  scaledMatrix.row (0).cross (scaledMatrix.row (2)),
+                  scaledMatrix.row (1).cross (scaledMatrix.row (2));
+
+  // expression template
+  const auto len = crossProduct.rowwise ().norm ();
+  const auto normalized = crossProduct.rowwise ().normalized ();
```

Comment:
In theory, we only need to normalize the vector that is returned return. Second thing, be careful with the use of `auto` and Eigen types. https://eigen.tuxfamily.org/dox/TopicPitfalls.html

Reply:
Yeah, I modified the last auto to `Scalar`

Reply:
My comment pertained line 280
```cpp
crossProduct.rowwise ().normalized ()
```

The method returns a `const MatrixBase< Derived >::PlainObject` object type. [ref](https://eigen.tuxfamily.org/dox/classEigen_1_1MatrixBase.html#a5cf2fd4c57e59604fd4116158fd34308)

Reply:
Wouldn't `rowwise` result in [this method](https://eigen.tuxfamily.org/dox/classEigen_1_1VectorwiseOp.html#a56b1ecf2069f3470f285360d05f9f774) being called which returns an expression not a `PlainObject`?

Reply:
You are right. But in what manner does that change my claim that you should not use `auto` here?

Reply:
So I rewrote it to remove the nearly useless statement with auto. The code wasn't required. I used auto to keep the expression template (ie no computation) without having to write the expression. Since it wasn't a temporary there wasn't an issue. But it's gone now and the code looks better.